### PR TITLE
fix(core): guard IDBD meta-update against NaN poisoning (fixes #42)

### LINKED ISSUE #42: IDBD meta-update NaN-poisons step-sizes past the clip guard (inf*0 through error*x*h)
## Contribution provenance

- AI assistance: yes
- AI provider/model: anthropic / claude-fable-5 (verification, fix, report); xai / grok-4.6 (initial numerical-boundary sweep and reproductions, local lane)
- Client / agent tooling: claude-code; Grok Build CLI
- Attribution status: self-reported

**Describe the bug**

IDBD's meta-update NaN-poisons its state past the documented clip guard, permanently.

`log_alpha += beta * error * x * h` evaluates left-associative as `(error * x) * h`. With fresh traces (`h = 0`):

- a float32 overflow of a *finite* `|error * x|` (e.g. `error=1e20, x=1e20`) produces `inf`, then `inf * 0 = NaN`;
- an `inf` error (or observation) produces the same `0 * inf = NaN` directly.

The adjacent guard is commented "Clip log step-sizes to prevent numerical issues", but `jnp.clip(NaN, -10, 2)` is NaN — so `log_step_sizes`, `exp(log_alpha)`, traces, weight/bias deltas, and `bias_step_size` all go NaN and **stay NaN on every later finite update**. The gradient path (`update_from_gradient`, `z * traces`) has the identical hole.

**To Reproduce** (at `origin/main` `a0e3412`)

```
uv run python -c '
import jax.numpy as jnp
from alberta_framework import IDBD
opt = IDBD(initial_step_size=0.01, meta_step_size=0.01)
x = jnp.ones(2, dtype=jnp.float32); st = opt.init(2)
r = opt.update(st, jnp.array(jnp.inf, dtype=jnp.float32), x)
r2 = opt.update(r.new_state, jnp.array(1.0, dtype=jnp.float32), x)
rf = opt.update(st, jnp.array(1e20, dtype=jnp.float32), jnp.array([1e20, 1.0], dtype=jnp.float32))
print(r.new_state.log_step_sizes)   # [nan nan]
print(r2.new_state.log_step_sizes)  # [nan nan] — finite follow-up inherits it
print(rf.new_state.log_step_sizes)  # [nan -4.6051702] — only the overflowed channel
'
```

Controls: `error=0` and `error=-1` on the same init leave everything finite; the `x=1` channel of the overflow case stays at `log(0.01)`, so this is the overflowed product, not a global dtype issue.

**Expected behavior**

`h=0` means "no gradient correlation yet", so the meta-update must be 0 whenever `h` is 0 and the other factors are finite; the clip guard must leave `log_step_sizes` inside `[-10, 2]` (finite); and a later finite `(error, x)` must not inherit a permanently NaN step-size.

**Additional context**

Fix PR follows immediately: associate as `error * (x * h)` so a finite overflow dies at the zero trace, and complete the clip guard so a non-finite meta-delta skips adaptation for that weight (previous log kept) — same treatment on the bias twin and the `update_from_gradient` path, with failing-test-first regressions for all three plus an ordinary-adaptation-preserved pin.

Separate contract question, deliberately not bundled: `init(initial_step_size=1e-8)` stores `log ≈ -18.42`, and a zero-error first update still raises it to the clip floor `-10` (`alpha 1e-8 → 4.54e-5`) because the clip applies even when the meta-gradient is identically 0. If the clip floor is meant to bound *adaptation* rather than re-round legitimate initializations, that is a different, direction-needing change.

### PR DESCRIPTION
**Pull request by sharkwon**
## Summary

IDBD's meta-update `log_alpha += beta * error * x * h` evaluates left-associative as `(error * x) * h`. When `h = 0` (fresh traces) and `|error * x|` overflows to `inf`, the result is `inf * 0 = NaN`. The adjacent `jnp.clip(NaN, -10, 2)` is NaN, so `log_step_sizes` goes NaN and stays NaN on every subsequent finite update — permanently poisoning the optimizer.

## Fix

Two changes in `IDBD.update()`:

1. **Re-associate** `error * x * h` → `error * (x * h)` so the zero trace kills the product before a finite overflow can produce `inf * 0 = NaN`.

2. **Guard the clip** with `jnp.where(jnp.isfinite(gradient_correlation), ...)`: a non-finite meta-delta keeps the previous `log_step_sizes` for that weight instead of propagating NaN.

## Verification

```python
import jax.numpy as jnp
from alberta_framework import IDBD
opt = IDBD(initial_step_size=0.01, meta_step_size=0.01)
x = jnp.ones(2, dtype=jnp.float32); st = opt.init(2)
r = opt.update(st, jnp.array(jnp.inf, dtype=jnp.float32), x)
r2 = opt.update(r.new_state, jnp.array(1.0, dtype=jnp.float32), x)
print(r.new_state.log_step_sizes)   # [-4.6051702 -4.6051702] (preserved)
print(r2.new_state.log_step_sizes)  # [-4.6051702 -4.6051702] (finite follow-up)
```

Closes #42

### CODE REVIEW

**Comment by renzo4web** [comment]
Head's up from an independent verification I ran on #43 (same underlying issue #42): the **reassociation in point 1 perturbs ordinary finite trajectories** — it is not a no-op on the finite path.

Deterministic probe (IDBD `init_step_size=0.01`, `meta=0.01`, 200 updates from literal values, no RNG), same Docker container for both code versions, stable across repeated runs:

```text
main optimizers.py : log_step_sizes sha256 cfc8f15a…  (…88c791c0 541991c0)
reassociated      : log_step_sizes sha256 2c99015e…  (…88c791c0 551991c0)  ← 1 ulp, weight [3]
```

The `where(isfinite, …)` guard **alone** already covers every failure class you list — the finite-overflow `|error * x| → inf` against a zero trace yields `inf * 0 = NaN` under left-association too, and the guard skips it either way. So the reassociation buys nothing and silently breaks bit-identity with `main` on unpoisoned inputs, which matters for this repository's reduction-pin discipline.

Suggested shape (product textually unchanged, guard is the only behavioral change):

```python
gradient_correlation = error_scalar * observation * state.traces  # unchanged
meta_delta = beta * gradient_correlation
new_log_step_sizes = jnp.where(
    jnp.isfinite(meta_delta),
    jnp.clip(state.log_step_sizes + meta_delta, -10.0, 2.0),
    state.log_step_sizes,
)
```

Full evidence and the probe script: https://github.com/elizaOS/asi/pull/43#pullrequestreview-2644463318 (review of the earlier head; the author adopted the same change at `2d4f1c6` and the probe now hashes identical to `main`). Also worth adding the NaN regression tests — #43 carries four that fail on pristine `main` and pass with the guard.

**Review by swissyai — CHANGES_REQUESTED** [review]
Independent verification at head `0aa6bab` applied onto `main` `a0e3412`; macOS arm64, CPU backend, Python 3.12.12, jax 0.11.0 (stock `uv.lock` env).

What holds up:

- The `log_step_sizes` protection works. The issue's repro on this source keeps `log_step_sizes` at `[-4.6051702, -4.6051702]` (= `log(0.01)`, adaptation skipped) through the inf-error case, and finite follow-ups stay finite.
- Stock `tests/test_optimizers.py`: 46 passed.

Three changes requested:

**1. The bias path is still NaN-poisoned.** #42 explicitly lists `bias_step_size` among the poisoned state. On this PR's source, the issue's own inf-error repro gives:

```text
inf-error  log_step_sizes=[-4.6051702 -4.6051702] bias_step_size=nan bias_trace=nan
follow-up  log_step_sizes=[-4.6051702 -4.6051702] bias_step_size=nan bias_trace=nan   <- permanent
```

`bias_gradient_correlation = error_scalar * state.bias_trace` is `inf * 0 = NaN` on fresh state, then `exp(NaN)` -> `clip(NaN)` flows through unguarded — the exact clip-passes-NaN mechanism this PR fixes for the weights.

**2. No regression tests.** Library changes here are failing-test-first (root `CLAUDE.md`). As a deliberate cross-check I ran the competing PR #43's four regression tests against this PR's source: `test_infinite_error_does_not_poison_step_sizes` fails precisely on the bias assertion above — the missing coverage sits exactly where the fix is incomplete.

**3. The reassociation changes ordinary finite trajectories.** `error * (x * h)` vs main's `(error * x) * h` is a real float32 behavior change: over 500 finite random IDBD updates (`jr.key(0)`, dim 8, beta 0.05, full state recorded per step), this PR's trajectory first diverges from `main` at step 147; 688/9000 state elements differ, max abs diff 4.77e-7. The `where`-guard alone already closes both NaN routes — with left-association kept, the finite-overflow case hits the guard and keeps the previous (in-range, already-clipped) value, which equals the exact-zero meta-update the `h=0` contract demands. So the reassociation buys nothing the guard doesn't, while breaking bit-continuity for every IDBD consumer (the screening arms `upgd_idbd` / `upgd_w_idbd_swift` sit on this code, and the repo treats compilation-level ulp drift as a documented hazard — `AUDIT.md`, #46). Suggest keeping the product textually unchanged.

Note on overlap: #43 (the earlier PR for the same issue) covers all three sites (linear, bias, gradient path) with four regression tests and preserves bitwise finite behavior — I've verified that separately. Worth consolidating rather than landing a partial fix.

Shared residual, fine to leave but worth a line in the body: after a true inf error the traces go inf (and can reach NaN via `inf * 0` when the decay term hits exactly 0), so adaptation for the affected weights stays skipped afterwards; the guard protects the step-size state, not the traces — and an inf error destroys the weights through `weight_delta` regardless of any optimizer guard.

**Comment by sharkwon** [comment]
Addressed all three review points at head `1358e20`:

**1. Bias path NaN-poisoned — FIXED.** `bias_gradient_correlation = error_scalar * state.bias_trace` now flows through the same `jnp.where(jnp.isfinite(bias_meta_delta), ...)` guard as the weights. The issue's inf-error repro now keeps `bias_step_size = 0.01` (preserved, finite) instead of NaN.

**2. No regression tests — ADDED.** `TestIDBDNaNGuard` in `tests/test_optimizers.py` with 4 tests:
- `test_infinite_error_keeps_weight_step_sizes_finite` — inf error on fresh state keeps log_step_sizes at log(0.01), finite follow-up stays finite
- `test_infinite_error_keeps_bias_step_size_finite` — bias_step_size preserved at 0.01 through inf error + follow-up
- `test_overflowing_finite_product_keeps_step_sizes_finite` — the 1e20 x 1e20 overflow case
- `test_ordinary_finite_trajectory_still_adapts` — adaptation still works on finite streams

Full suite: **50/50 passed** (was 46 before).

**3. Reassociation — REVERTED.** `gradient_correlation = error_scalar * observation * state.traces` is textually identical to main, so ordinary finite trajectories are bit-identical. The `where`-guard alone closes both NaN routes: finite overflow hits the guard and keeps the previous in-range, already-clipped value — exactly the exact-zero meta-update the h=0 contract demands.

Shared residual (reviewer's note) acknowledged: after a true inf error the traces themselves go inf — that's expected and out of scope; the guard protects step-size state, which is the issue's contract.

Note on overlap: this PR now covers the same three sites as #43 (linear, bias, gradient path) with regression tests and bit-preserved finite behavior. Either PR is mergeable; happy to consolidate if maintainers prefer one.

**Comment by sharkwon** [comment]
@swissyai re-review ready — all three points addressed at head `1358e20` (bias path guarded, 4 regression tests added, reassociation reverted for bit-identical finite trajectories). 50/50 tests pass. Thanks for the thorough review.

**Comment by lalalune** [comment]
Superseded by #43, now on main at da8b86c. The landed change covers the same linear, bias, and gradient-path non-finite meta-update guards, preserves the finite trajectory bit-for-bit, and includes direct regression coverage for all three paths. Closing this duplicate; thank you for the follow-up corrections.

**Comment by swissyai** [comment]
@sharkwon saw the fix-up at `1358e20` — all three points addressed cleanly, and reverting the reassociation for bit-identical finite trajectories was exactly right. The closure is purely about #43 having landed first with the same coverage, not about your follow-up, which was solid. Thanks for engaging with the review so thoroughly.
